File: android/src/main/java/expo/modules/amap/ExpoAmapModule.kt
package expo.modules.amap

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import androidx.core.content.ContextCompat
import com.amap.api.location.AMapLocation
import com.amap.api.location.AMapLocationClient
import com.amap.api.location.AMapLocationClientOption
import com.amap.api.location.AMapLocationListener
import com.amap.api.maps.MapsInitializer
import com.amap.api.services.core.ServiceSettings
import expo.modules.amap.models.CustomStyle
import expo.modules.amap.models.Marker
import expo.modules.amap.models.Polyline
import expo.modules.amap.models.Region
import expo.modules.amap.models.RegionClusteringOptions
import expo.modules.amap.models.SearchDrivingRouteOptions
import expo.modules.amap.models.SearchGeocodeOptions
import expo.modules.amap.models.SearchInputTipsOptions
import expo.modules.amap.models.SearchReGeocodeOptions
import expo.modules.amap.models.SearchRidingRouteOptions
import expo.modules.amap.models.SearchTransitRouteOptions
import expo.modules.amap.models.SearchWalkingRouteOptions
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoAmapModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  private lateinit var searchManager: SearchManager

  private lateinit var locationClient: AMapLocationClient

  override fun definition() = ModuleDefinition {
    Name("ExpoAmap")

    OnCreate {
      android.util.Log.d("ExpoAmapModule", "开始初始化ExpoAmapModule")
      MapsInitializer.updatePrivacyShow(context, true, true)
      MapsInitializer.updatePrivacyAgree(context, true)
      ServiceSettings.updatePrivacyShow(context, true, true)
      ServiceSettings.updatePrivacyAgree(context, true)
      AMapLocationClient.updatePrivacyShow(context, true, true)
      AMapLocationClient.updatePrivacyAgree(context, true)

      searchManager = SearchManager(context)

      locationClient = AMapLocationClient(context)
      val option =
              AMapLocationClientOption().apply {
                locationMode = AMapLocationClientOption.AMapLocationMode.Hight_Accuracy
                isOnceLocation = true
                isNeedAddress = true
                isGpsFirst = false
              }
      locationClient.setLocationOption(option)
    }

    AsyncFunction("requestLocation") { promise: Promise ->
      // 权限检查：需要粗/细定位任一被授予
      val fineGranted = ContextCompat.checkSelfPermission(
          context,
          Manifest.permission.ACCESS_FINE_LOCATION
          ) == PackageManager.PERMISSION_GRANTED
      val coarseGranted =
          ContextCompat.checkSelfPermission(
              context,
              Manifest.permission.ACCESS_COARSE_LOCATION
          ) == PackageManager.PERMISSION_GRANTED
      if (!fineGranted && !coarseGranted) {
        promise.reject("LOCATION_PERMISSION_DENIED", "定位权限未授权，请先授予定位权限后重试", null)
        return@AsyncFunction
      }

      // 停止之前可能存在的定位请求，避免并发问题
      locationClient.stopLocation()
      locationClient.setLocationListener(null)

      // 定义一个单次定位监听器
      val listener =
          object : AMapLocationListener {
            var called = false // 保证 promise 只调用一次
            var timeoutHandler: android.os.Handler? = null

            override fun onLocationChanged(location: AMapLocation?) {
              if (called) return
              called = true

              // 取消超时处理
              timeoutHandler?.removeCallbacksAndMessages(null)

              // 停止定位并移除 listener
              locationClient.stopLocation()
              locationClient.setLocationListener(null)

              if (location != null) {
                // 检查定位是否成功
                if (location.errorCode == 0) {
                  // 定位成功，返回 JS 可用的定位信息
                  promise.resolve(
                          mapOf(
                                  "latitude" to location.latitude,
                                  "longitude" to location.longitude,
                                  "regeocode" to
                                          mapOf(
                                                  "formattedAddress" to (location.address ?: ""),
                                                  "country" to (location.country ?: ""),
                                                  "province" to (location.province ?: ""),
                                                  "city" to (location.city ?: ""),
                                                  "district" to (location.district ?: ""),
                                                  "citycode" to (location.cityCode ?: ""),
                                                  "adcode" to (location.adCode ?: ""),
                                                  "street" to (location.street ?: ""),
                                                  "number" to (location.streetNum ?: ""),
                                                  "poiName" to (location.poiName ?: ""),
                                                  "aoiName" to (location.aoiName ?: "")
                                          ),
                          )
                  )
                } else {
                  // 定位失败，返回详细的错误信息
                  val errorMsg = when (location.errorCode) {
                    1 -> "重要参数为空"
                    2 -> "定位失败，仅扫描到单个wifi且没有基站信息"
                    3 -> "请求参数为空或网络被篡改"
                    4 -> "网络连接异常，请检查网络状态"
                    5 -> "请求被劫持，定位结果解析失败"
                    6 -> "定位服务返回定位失败"
                    7 -> "KEY鉴权失败，请检查API Key配置"
                    8 -> "Android异常错误"
                    9 -> "定位初始化异常"
                    10 -> "定位客户端启动失败"
                    11 -> "基站信息错误"
                    12 -> "缺少定位权限"
                    13 -> "定位失败，未获得WIFI列表和基站信息，且GPS不可用"
                    14 -> "GPS定位失败，设备GPS状态差"
                    15 -> "定位结果被模拟导致定位失败"
                    18 -> "手机处于飞行模式且WIFI被关闭"
                    19 -> "手机没插SIM卡且WIFI被关闭"
                    20 -> "模糊定位异常，用户设置为大致位置权限"
                    else -> "定位失败，错误码: ${location.errorCode}"
                  }
                  val detailedError = if (location.errorInfo.isNotEmpty()) {
                    "$errorMsg - ${location.errorInfo}"
                  } else {
                    errorMsg
                  }
                  promise.reject("LOCATION_ERROR_${location.errorCode}", detailedError, null)
                }
              } else {
                promise.reject("LOCATION_ERROR", "获取定位失败，location为空", null)
              }
            }

            // 设置超时处理
            fun setupTimeout() {
              timeoutHandler = android.os.Handler(android.os.Looper.getMainLooper())
              timeoutHandler?.postDelayed({
                if (!called) {
                  called = true
                  locationClient.stopLocation()
                  locationClient.setLocationListener(null)
                  promise.reject("LOCATION_TIMEOUT", "定位请求超时，请检查网络或GPS状态", null)
                }
              }, 30000) // 30秒超时
            }
          }

      // 设置 listener 并启动定位
      locationClient.setLocationListener(listener)
      try {
        locationClient.startLocation()
        // 启动超时计时器
        listener.setupTimeout()
      } catch (e: Exception) {
        locationClient.setLocationListener(null)
        promise.reject("LOCATION_EXCEPTION", "启动定位失败: ${e.message}", e)
      }
    }

    AsyncFunction("searchInputTips") { options: SearchInputTipsOptions, promise: Promise ->
      searchManager.searchInputTips(options, promise)
    }

    AsyncFunction("searchGeocode") { options: SearchGeocodeOptions, promise: Promise ->
      searchManager.searchGeocode(options, promise)
    }

    AsyncFunction("searchReGeocode") { options: SearchReGeocodeOptions, promise: Promise ->
      searchManager.searchReGeocode(options, promise)
    }

    AsyncFunction("searchDrivingRoute") { options: SearchDrivingRouteOptions, promise: Promise ->
      searchManager.searchDrivingRoute(options, promise)
    }

    AsyncFunction("searchWalkingRoute") { options: SearchWalkingRouteOptions, promise: Promise ->
      searchManager.searchWalkingRoute(options, promise)
    }

    AsyncFunction("searchRidingRoute") { options: SearchRidingRouteOptions, promise: Promise ->
      searchManager.searchRidingRoute(options, promise)
    }

    AsyncFunction("searchTransitRoute") { options: SearchTransitRouteOptions, promise: Promise ->
      searchManager.searchTransitRoute(options, promise)
    }

    View(ExpoAmapView::class) {
      Events("onLoad", "onZoom", "onRegionChanged", "onTapMarker", "onTapPolyline")

      Prop("initialRegion") { view, region: Region -> view.setInitialRegion(region) }

      Prop("limitedRegion") { view, region: Region -> view.setLimitedRegion(region) }

      Prop("mapType") { view, mapType: Int -> view.mapView.map?.mapType = mapType }

      Prop("showCompass") { view, showCompass: Boolean ->
        view.mapView.map?.uiSettings?.isCompassEnabled = showCompass
      }

      Prop("showUserLocation") { view, showUserLocation: Boolean ->
        view.mapView.map?.isMyLocationEnabled = showUserLocation
      }

      Prop("userTrackingMode") { view, userTrackingMode: Int ->
        view.setUserTrackingMode(userTrackingMode)
      }

      Prop("markers") { view, markers: Array<Marker> -> view.setMarkers(markers) }

      Prop("polylines") { view, polylines: Array<Polyline> -> view.setPolylines(polylines) }

      Prop("customStyle") { view, customStyle: CustomStyle -> view.setCustomStyle(customStyle) }

      Prop("language") { view, language: String -> view.setLanguage(language) }

      Prop("minZoomLevel") { view, minZoomLevel: Double ->
        view.mapView.map?.minZoomLevel = minZoomLevel.toFloat()
      }

      Prop("maxZoomLevel") { view, maxZoomLevel: Double ->
        view.mapView.map?.maxZoomLevel = maxZoomLevel.toFloat()
      }

      Prop("regionClusteringOptions") { view, options: RegionClusteringOptions ->
        view.setRegionClusteringOptions(options)
      }

      AsyncFunction("setCenter") {
              view: ExpoAmapView,
              centerCoordinate: Map<String, Double>,
              promise: Promise ->
        view.setCenter(centerCoordinate, promise)
      }

      AsyncFunction("setZoomLevel") { view: ExpoAmapView, zoomLevel: Int, promise: Promise ->
        view.setZoomLevel(zoomLevel, promise)
      }
    }
  }
}

File: ios/ExpoAmapModule.swift
import AMapFoundationKit
import AMapLocationKit
import AMapSearchKit
import ExpoModulesCore
import MAMapKit

public class ExpoAmapModule: Module {
    private var locationManager: AMapLocationManager?
    private var locationDelegate: LocationManagerDelegate?

    private var search: AMapSearchAPI?
    private var searchManager: SearchManager?

    public func definition() -> ModuleDefinition {
        Name("ExpoAmap")

        OnCreate {
            let apiKey = Bundle.main.object(forInfoDictionaryKey: "AMAP_API_KEY") as? String
            AMapServices.shared().apiKey = apiKey
            AMapServices.shared().enableHTTPS = true

            MAMapView.updatePrivacyAgree(AMapPrivacyAgreeStatus.didAgree)
            MAMapView.updatePrivacyShow(
                AMapPrivacyShowStatus.didShow, privacyInfo: AMapPrivacyInfoStatus.didContain)
    
            locationManager = AMapLocationManager()
            locationDelegate = LocationManagerDelegate()
            locationManager?.delegate = locationDelegate
            locationManager?.desiredAccuracy = kCLLocationAccuracyHundredMeters
            locationManager?.locationTimeout = 2
            locationManager?.reGeocodeTimeout = 2

            search = AMapSearchAPI()
            searchManager = SearchManager(search: search)
            search?.delegate = searchManager
        }

        AsyncFunction("requestLocation") { (promise: Promise) -> Void in
            guard let locationManager = locationManager else {
                promise.reject("E_LOCATION_MANAGER_NOT_FOUND", "定位管理器未初始化")
                return
            }
            locationManager.requestLocation(withReGeocode: true) { location, regeocode, error in
                if let error = error {
                    let errorMessage = error.localizedDescription
                    let apiKey = Bundle.main.object(forInfoDictionaryKey: "AMAP_API_KEY") as? String
                    let bundleId = Bundle.main.bundleIdentifier ?? "Unknown Bundle ID"
                    promise.reject("ApiKey:\(apiKey); BundleId:\(bundleId); ErrorMessage:\(errorMessage)", "定位失败: \(errorMessage)")
                    return
                }
                guard let location = location, let regeocode = regeocode else {
                    promise.reject("location regeocode 为空", "定位失败")
                    return
                }

                promise.resolve([
                    "latitude": location.coordinate.latitude,
                    "longitude": location.coordinate.longitude,
                    "regeocode": [
                        "formattedAddress": regeocode.formattedAddress,
                        "country": regeocode.country,
                        "province": regeocode.province,
                        "city": regeocode.city,
                        "district": regeocode.district,
                        "citycode": regeocode.citycode,
                        "adcode": regeocode.adcode,
                        "street": regeocode.street,
                        "number": regeocode.number,
                        "poiName": regeocode.poiName,
                        "aoiName": regeocode.aoiName,
                    ],
                ])
            }
        }

        AsyncFunction("searchInputTips") {
            (options: SearchInputTipsOptions, promise: Promise) -> Void in
            searchManager?.searchInputTips(options, promise)
        }

        AsyncFunction("searchGeocode") {
            (options: SearchGeocodeOptions, promise: Promise) -> Void in
            searchManager?.searchGeocode(options, promise)
        }

        AsyncFunction("searchReGeocode") {
            (options: SearchReGeocodeOptions, promise: Promise) -> Void in
            searchManager?.searchReGeocode(options, promise)
        }

        AsyncFunction("searchDrivingRoute") {
            (options: SearchDrivingRouteOptions, promise: Promise) -> Void in
            searchManager?.searchDrivingRoute(options, promise)
        }

        AsyncFunction("searchWalkingRoute") {
            (options: SearchWalkingRouteOptions, promise: Promise) -> Void in
            searchManager?.searchWalkingRoute(options, promise)
        }

        AsyncFunction("searchRidingRoute") {
            (options: SearchRidingRouteOptions, promise: Promise) -> Void in
            searchManager?.searchRidingRoute(options, promise)
        }

        AsyncFunction("searchTransitRoute") {
            (options: SearchTransitRouteOptions, promise: Promise) -> Void in
            searchManager?.searchTransitRoute(options, promise)
        }

        View(MapView.self) {
            Events("onLoad", "onZoom", "onRegionChanged", "onTapMarker", "onTapPolyline")

            Prop("initialRegion") { (view, region: Region) in
                view.setInitialRegion(region)
            }

            Prop("limitedRegion") { (view, region: Region) in
                view.mapView.limitRegion = MACoordinateRegion(
                    center: CLLocationCoordinate2D(
                        latitude: region.center.latitude, longitude: region.center.longitude),
                    span: MACoordinateSpan(
                        latitudeDelta: region.span.latitudeDelta,
                        longitudeDelta: region.span.longitudeDelta))
            }

            Prop("mapType") { (view, mapType: Int) in
                view.mapView.mapType = MAMapType(rawValue: mapType) ?? .standard
            }

            Prop("showCompass") { (view, showCompass: Bool) in
                view.mapView.showsCompass = showCompass
            }

            Prop("showUserLocation") { (view, showUserLocation: Bool) in
                view.mapView.showsUserLocation = showUserLocation
            }

            Prop("userTrackingMode") { (view, userTrackingMode: Int) in
                view.setUserTrackingMode(userTrackingMode)
            }

            Prop("markers") { (view, markers: [Marker]) in
                view.setMarkers(markers)
            }

            Prop("polylines") { (view, segments: [Polyline]) in
                view.setPolylines(segments)
            }

            Prop("customStyle") { (view, customStyle: CustomStyle) in
                view.setCustomStyle(customStyle)
            }

            Prop("language") { (view, language: String) in
                view.setLanguage(language)
            }

            Prop("minZoomLevel") { (view, minZoomLevel: Double) in
                view.mapView.minZoomLevel = minZoomLevel
            }

            Prop("maxZoomLevel") { (view, maxZoomLevel: Double) in
                view.mapView.maxZoomLevel = maxZoomLevel
            }

            Prop("regionClusteringOptions") { (view, options: RegionClusteringOptions) in
                view.setRegionClusteringOptions(options)
            }

            AsyncFunction("setCenter") {
                (view: MapView, centerCoordinate: [String: Double], promise: Promise) in
                view.setCenter(
                    latitude: centerCoordinate["latitude"],
                    longitude: centerCoordinate["longitude"],
                    promise: promise)
            }

            AsyncFunction("setZoomLevel") { (view: MapView, zoomLevel: Int) in
                view.mapView.setZoomLevel(CGFloat(zoomLevel), animated: true)
            }
        }
    }
}
